### PR TITLE
Alertmanager alerts limits

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR adds new limits in Alertmanager:
- total number of alerts that a user can have in Alertmanager's memory (`-alertmanager.max-alerts-count`)
- total size of alerts that user can have in Alertmanager's memory (`-alertmanager.max-alerts-size-bytes`)

These are overrideable per tenant. When limits are reached, Alertmanager will reject more alerts, produce a log message and increment `cortex_alertmanager_insert_alert_failures_total` metric. Additional metric to track behaviour of alerts limiter are: `cortex_alertmanager_alerts_limiter_current_alerts_count` and `cortex_alertmanager_alerts_limiter_current_alerts_size_bytes`.

This PR builds on top of https://github.com/cortexproject/cortex/pull/4237, and will be rebased once https://github.com/cortexproject/cortex/pull/4237 is merged.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pstibrany** [comment]
Marking as draft until https://github.com/cortexproject/cortex/pull/4237 is merged.

**Comment by gotjosh** [comment]
Can you assign this to me? I'll give you a first pass on it.

**Comment by pstibrany** [comment]
> Can you assign this to me? I'll give you a first pass on it.

Unfortunately assigning only cortex-project members :( But don't let that stop you. Thank you!

**Comment by pstibrany** [comment]
Rebased on top of master now, it's ready for review.

**Comment by stevesg** [comment]
Will let Josh do a first pass and come back to this one.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [inline pkg/alertmanager/alertmanager.go:655]: An alert is considered unique based on the hash of its labels (fingerprint), technically if an alert comes in with a size that's OK, then its annotations change and its size is no longer acceptable we would bypass it because we would consider its size to be OK because it was already there.

I believe we need to check that if it does exist, its new size would not tip us over the limit. Unless you want the semantic to be: **If we had the alert before, we'll always accept it regardless of whenever it tips us over the limit**

If this is the case, can we add a comment?

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [inline pkg/alertmanager/alertmanager.go:680]: nit: you can do this before taking the lock to minimise lock time

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [inline pkg/alertmanager/alertmanager.go:698]: ditto from above

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [inline pkg/alertmanager/alertmanager.go:621]:
```suggestion
// alertsLimiter helps us limit the number and size of alerts being received by the Alertmanager.
// We consider an alert unique based on its fingerprint (a hash of its labels) and its size it's determined by the sum of bytes of its labels, annotations, and GeneratorURL.
type alertsLimiter struct {
```

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [inline pkg/alertmanager/alertmanager.go:721]:
```suggestion
        // We include the generatorURL as part of the alert size given it includes the user-input query of the alert itself.
	size += len(alert.GeneratorURL)
```

Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:721]: I'm mostly looking for a discussion point here (not saying it is wrong), I understand this is user input but a) we already have a limit for the size of the ruler configuration to catch egregiousness and b) it seems a bit unfair that we take a chunk of their limit based on something that might be beyond their control (they can reduce their query length but this also includes the URL of where it's going to)

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [inline pkg/alertmanager/alertmanager.go:248]:
```suggestion
			Name: "alertmanager_alerts_insert_failures_total",
```

Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:248]: To keep the `alertmanager_alerts` prefix homogenous?

File: pkg/util/validation/limits.go
Comment (gotjosh) [inline pkg/util/validation/limits.go:186]:
```suggestion
	f.IntVar(&l.AlertmanagerMaxAlertsCount, "alertmanager.max-alerts-count", 0, "Maximum number of alerts that a single user can have. Inserting more alerts will fail with a log message and metric increment. 0 = no limit.")
```

Comment (gotjosh) [reply pkg/util/validation/limits.go:186]: this typo is repeated across several places including below and the changelog.

File: pkg/util/validation/limits.go
Comment (gotjosh) [inline pkg/util/validation/limits.go:187]:
```suggestion
	f.IntVar(&l.AlertmanagerMaxAlertsSizeBytes, "alertmanager.max-alerts-size-bytes", 0, "Maximum total size of alerts that single a user can have, alert size is the sum of the bytes of its labels, annotations and generatorURL. Inserting more alerts will fail with a log message and metric increment. 0 = no limit.")
```

File: pkg/alertmanager/multitenant.go
Comment (gotjosh) [inline pkg/alertmanager/multitenant.go:227]:
```suggestion
	// Size of the alert is computed from alert labels, annotations, and generatorURL.
```

**Review by gotjosh — COMMENTED** [review]
🎖️ Fantastic job @pstibrany ! Most of my comments are nits and questions (for the benefit of my understanding).

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:655]: > If we had the alert before, we'll always accept it regardless of whenever it tips us over the limit

That was indeed my intention, mostly to make sure that we don't accidentally reject "resolved" alerts. But if the size has changed, it may make sense to reject it if it's over limit. I don't have strong opinion on this.

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:680]: Thanks. Indeed, that was my plan, and I failed to follow it :) Let me fix it.

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:721]: I think the alerts size limit (say 10 MiB) should be high enough that relatively short (< 50 chars) URL will not make a big difference.

I think including GeneratorURL makes sense because it better reflects size of the alert, if we do, but I see your point that user cannot control it, and I'd be fine removing it from the computation too.

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:248]: sounds good

File: pkg/util/validation/limits.go
Comment (pstibrany) [reply pkg/util/validation/limits.go:186]: not a typo, just poor grammar skills. thanks, will fix :)

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:658]: Also ... when new alert arrives and "merges" with existing one, it cannot update labels (fingerprint would change) or annotations (right now, at least).

See the merging code: https://github.com/cortexproject/cortex/blob/5e496f428302a195a68cc287d97641561dbe0afd/vendor/github.com/prometheus/alertmanager/types/types.go#L366

File: pkg/alertmanager/alertmanager_metrics.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager_metrics.go:216]: Should we call it `cortex_alertmanager_alerts_insert_rate_limited_total` similarly to `alertmanager_notification_rate_limited_total`?

File: pkg/alertmanager/alertmanager_metrics.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager_metrics.go:220]: [nit] Gauges typically don't have the `_count` suffix.

```suggestion
			"cortex_alertmanager_alerts_limiter_current_alerts",
```

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager.go:250]: This has reason but when we expose it via `cortex_alertmanager_alerts_insert_failures_total ` we loose it.

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager.go:255]: [nit, design] Why not passing the registered to `newAlertsLimiter` and register metrics there? I think it may lead to cleaner code.

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager.go:661]: If we don't have to do anything in case it already exists, then we can move this check at the top of the function and early `return nil`.

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager.go:718]: Have you considered `unsafe.Sizeof()` ?

**Review by pracucci — COMMENTED** [review]
Good job! I left a couple of comments I would be glad if you could take a look 🙏 🙇

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:728]: I do agree that it better reflects the size of the alert, let's leave it in. It's already documented as part of the flag which was my main concern.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:661]: > or annotations (right now, at least)

I see that is true for the labels, which is expected but is it also for the annotations? I can't seem to see it in the code - am I missing something?

Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:661]: > I see that is true for the labels, which is expected but is it also for the annotations? I can't seem to see it in the code - am I missing something?

No, you’re right. I misread the code I linked… line 372 is copying everything over, including annotations.

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:718]: `unsafe.Sizeof` doesn’t return correct size for map, we still need to count size of keys/values manually.

File: pkg/alertmanager/alertmanager_metrics.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager_metrics.go:216]: `cortex_alertmanager_alerts_insert_rate_limited_total` is not strictly true... there is no rate limit here, just a limit for number of alerts or their size. What about `cortex_alertmanager_alerts_insert_limited_total` instead? (dropping "rate" part)

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:255]: You're right. I guess I had some reason to do it this way, but I don't see it anymore :)

File: pkg/alertmanager/alertmanager_metrics.go
Comment (pracucci) [reply pkg/alertmanager/alertmanager_metrics.go:216]: > What about cortex_alertmanager_alerts_insert_limited_total instead?

LGTM

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go]: I've removed the reason from "internal" metric. While it was nice to have, without exposing it to "aggregated" registry, it is useless, and we already expose current alert count and alerts size metrics to help troubleshooting.

**Comment by pstibrany** [comment]
Thank you for reviews, I've addressed all your comments. Please take a look again.

**Review by pracucci — APPROVED** [review]
LGTM (modulo a nit). Thanks a lot for addressing my feedback 🙏

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:652]: 🤦 you're right of course.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go]: Makes sense. In this case, I think rejection would be the best course of action - don't you think?

The alert that existed before should have a short enough `endsAt` that it would resolve itself eventually so there's little to no risk of just leaving the existing one to resolve by itself. On the contrary, accepting the alert that might tip us over the limit risks our availability.

Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go]: I think you're right, and I will make this change before merging.

Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go]: I've changed the logic such that existing alert that grows and doesn't fit the limit anymore if rejected.

**Pull request merged**
